File: simp-front-end/src/lib/types/recipe.ts
import exp from "constants";
import { UUID } from "crypto";

// export type Recipe = {
//   recipe_id: string;
//   title: string;
//   description?: string;
//   author_id?: string;
//   created_at: string;  // ❌ Not needed in `RecipeCreate`
//   ingredients: RecipeIngredient[];
//   steps: RecipeStep[];
//   images: RecipeImage[];
//   tags: Tag[];  // Backend expects UUID[], so we must transform this
//   favorited_by: RecipeFavorite[];  // ❌ Not needed in `RecipeCreate`
// };

// ✅ Define `RecipeCreate` correctly to match `RecipeCreateSchema`
// export type RecipeCreate = {
//   title: string;
//   description?: string;
//   author_id?: string;
//   ingredients: RecipeIngredientCreate[];
//   steps: RecipeStepCreate[];
//   images: RecipeImageCreate[];
//   tags: string[];
// };


export type RecipeLoad = {
  recipe_id: string;
  title: string;
};



export type RecipeUpdate = Partial<RecipeCreate>;

export type RecipeIngredient = {
  ingredient_id: string;
  ingredient_name: string;
  recipe_id: string;
  amount: number;
  measurement: string;
  created_at: string;
};

export type RecipeStep = {
  step_id: string;
  recipe_id: string;
  step_number: number;
  description: string;
  image_url?: string;
  created_at: string;
};

export type RecipeImage = {
  image_id: string;
  recipe_id: string;
  image_url: string;
  created_at: string;
};

export type Tag = {
  tag_id: string;
  name: string;
};

export type RecipeFavorite = {
  user_id: string;
  recipe_id: string;
  favorited_at: string;
};


// export type RecipeIngredientCreate = {
//   id: string; // Temporary ID for frontend sorting
//   ingredient_name: string;
//   amount: number;
//   measurement: string;
// };


// export type RecipeStepCreate = {
//   id: string; // Temporary ID for frontend sorting
//   step_number: number;
//   description: string;
// };

// export type RecipeImageCreate = {
//   id: string; // Temporary ID for frontend sorting
//   image_url: string;
// };



export type RecipeIngredientModify = {
  ingredient_id: string;
  ingredient_name: string;
  amount: number;
  measurement: string;
};


// export interface RecipeRetrieve {
//   title: string;
//   description?: string;
//   author_id?: string;
//   ingredients: RecipeIngredient[];
//   steps: RecipeStep[];
//   images: RecipeImage[];
//   tags: Tag[];
// }


// -------------------------------- New types.

// Retrieval Recipe

export type ListRecipe = {
  recipe_id: string;
  title: string;
  tags: string[];
  front_image: string;
  in_cauldron: boolean;
}

export type TagRetrieval = {
  tag_id: string;
  name: string;
}

// Create Recipes

export type RecipeCreate = {
  recipe_id?: string; 
  title: string;
  description: string;
  front_image: string;
  author_id: string;
  ingredients: RecipeIngredientCreate[];
  steps: RecipeStepCreate[]; 
  images: RecipeImageCreate[];
  tags: RecipeTagCreate[];
};


export type RecipeIngredientCreate = {
  ingredient_id?: string;
  ingredient_name: string;
  amount: number;
  measurement: string;
  position: number;
};

export type RecipeStepCreate = {
  step_number: number;
  description: string;
  image_url?: string;
};

export type RecipeImageCreate = {
  image_url: string;
}

export type RecipeTagCreate = {
  tag_id?: string;
  name: string;
}


  // RecipeAnalytics aggregates global analytics for a recipe (independent of a specific user).
  export type RecipeAnalytics = {
    recipe_analytics_id: string;
    recipe_id: string;
    minimum_price: number;
    total_calories?: number;
    updated_at: string;
    created_at: string;
  };
  
  // For creating a new recipe analytics record.
  export type RecipeAnalyticsCreate = {
    recipe_id: string;
    minimum_price: number;
    total_calories?: number;
  };
  
  // For updating an existing recipe analytics record.
  export type RecipeAnalyticsUpdate = Partial<RecipeAnalyticsCreate>;

File: simp-api-recipes/schemas/recipe.py
from pydantic import BaseModel
from typing import List, Optional
from uuid import UUID

class RecipeOut(BaseModel):
    recipe_id: UUID
    title: str
    front_image: str
    tags: List[str]
    in_cauldron: bool = False

    class Config:
        from_attributes = True

class CreateRecipeIngredient(BaseModel):
    ingredient_name: str
    amount: float
    measurement: str
    position: int

    class Config:
        from_attributes = True

class CreateRecipeStep(BaseModel):
    step_number: int
    description: str
    image_url: Optional[str] = None  # Now accepts null

    class Config:
        from_attributes = True

class CreateRecipeImage(BaseModel):
    image_url: str

    class Config:
        from_attributes = True

class CreateRecipeTag(BaseModel):
    tag_id: Optional[UUID] = None
    name: str

    class Config:
        from_attributes = True

class CreateRecipe(BaseModel):
    title: str
    description: str
    front_image: str
    author_id: UUID
    ingredients: List[CreateRecipeIngredient]
    steps: List[CreateRecipeStep] 
    images: List[CreateRecipeImage]
    tags: List[CreateRecipeTag]

    class Config:
        from_attributes = True

class EditRecipe(BaseModel):
    recipe_id: UUID
    title: str
    description: str
    front_image: str
    author_id: UUID
    ingredients: List[CreateRecipeIngredient]
    steps: List[CreateRecipeStep] 
    images: List[CreateRecipeImage]
    tags: List[CreateRecipeTag]

    class Config:
        from_attributes = True



class RetrieveTag(BaseModel):
    tag_id: UUID
    name: str

    class Config: 
        from_attributes = True


# class RecipeIngredientSchema(BaseModel):
#     ingredient_id: UUID
#     ingredient_name: str
#     amount: float
#     measurement: str

#     class Config:
#         from_attributes = True

# class RecipeStepSchema(BaseModel):
#     step_number: int
#     description: str
#     image_url: Optional[str] = None

#     class Config:
#         from_attributes = True
    

# class RecipeImageSchema(BaseModel):
#     image_url: str

#     class Config:
#         from_attributes = True

# class RecipeCreateSchema(BaseModel):
#     title: str
#     description: Optional[str] = None
#     author_id: Optional[UUID] = None
#     ingredients: List[RecipeIngredientSchema]
#     steps: List[RecipeStepSchema]
#     images: List[RecipeImageSchema]
#     tags: List[UUID]

#     class Config:
#         from_attributes = True

# class TagOut(BaseModel):
#     tag_id: UUID
#     name: str

#     class Config:
#         from_attributes = True

# class RecipeRetrieveSchema(BaseModel):
#     title: str
#     description: Optional[str] = None
#     author_id: Optional[UUID] = None
#     ingredients: List[RecipeIngredientSchema]
#     steps: List[RecipeStepSchema]
#     images: List[RecipeImageSchema]
#     tags: List[TagOut]

